norm_sq return type is complex

Hi, I am trying to use the `norm_sq` function on a tensor of complex numbers. Intuitively, it seems that the return value should be real. However, when I write something like this, the compilation fails.
Here's a sample program.
```cpp
#include <iostream>
#include <complex>
#include <xtensor/xtensor.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xnorm.hpp>
#include <xtensor/xcomplex.hpp>

using cpx = std::complex<double>;
using std::complex;

template<class T>
xt::xtensor<T, 1> mynorm(xt::xtensor<complex<T>, 2> &a)
{
  return xt::norm_sq(a, {0});
}

int main()
{
  xt::xtensor<cpx, 2> a = {
    { cpx(1, 1), cpx(-1, 1), cpx(-2, -2) },
    { cpx(-1, 0), cpx(0, 1), cpx(2, 2) }
  };

  // std::cout << a << std::endl;
  std::cout <<  mynorm(a) << std::endl;

}
```
This fails with a failed cast from `complex<double>` to `double`.
```
g++ -I /Users/scheibler/anaconda3/envs/mixiva/include ./test.cpp -o test_norm -std=c++14
In file included from ./test.cpp:4:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xtensor.hpp:18:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xbuffer_adaptor.hpp:20:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xstorage.hpp:21:
In file included from /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xtensor_simd.hpp:14:
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xutils.hpp:742:20: error: cannot convert 'std::__1::complex<double>' to 'double'
      without a conversion operator
            return static_cast<T>(std::forward<U>(u));
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xutils.hpp:757:16: note: in instantiation of function template specialization
      'xt::conditional_cast_functor<true, double>::operator()<std::__1::complex<double> >' requested here
        return conditional_cast_functor<condition, T>()(std::forward<U>(u));
               ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xassign.hpp:473:22: note: in instantiation of function template specialization
      'xt::conditional_cast<true, double, std::__1::complex<double> >' requested here
            *m_lhs = conditional_cast<is_narrowing, result_type>(*m_rhs);
                     ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xassign.hpp:319:22: note: in instantiation of member function
      'xt::stepper_assigner<xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 1, xt::layout_type::row_major,
      xt::xtensor_expression_tag>, xt::xreducer<xt::xreducer_functors<(lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), (lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> >, xt::layout_type::row_major>::run'
      requested here
            assigner.run();
                     ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xassign.hpp:328:20: note: in instantiation of function template specialization
      'xt::xexpression_assigner_base<xt::xtensor_expression_tag>::assign_data<xt::xtensor_container<xt::uvector<double,
      std::__1::allocator<double> >, 1, xt::layout_type::row_major, xt::xtensor_expression_tag>, xt::xreducer<xt::xreducer_functors<(lambda
      at /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), (lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> > >' requested here
        base_type::assign_data(e1, e2, trivial_broadcast);
                   ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xassign.hpp:201:40: note: in instantiation of function template specialization
      'xt::xexpression_assigner<xt::xtensor_expression_tag>::assign_xexpression<xt::xexpression<xt::xtensor_container<xt::uvector<double,
      std::__1::allocator<double> >, 1, xt::layout_type::row_major, xt::xtensor_expression_tag> >,
      xt::xexpression<xt::xreducer<xt::xreducer_functors<(lambda at /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5),
      (lambda at /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> > > >' requested here
            xexpression_assigner<tag>::assign_xexpression(e1, e2);
                                       ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xsemantic.hpp:587:13: note: in instantiation of function template specialization
      'xt::assign_xexpression<xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 1, xt::layout_type::row_major,
      xt::xtensor_expression_tag>, xt::xreducer<xt::xreducer_functors<(lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), (lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> > >' requested here
        xt::assign_xexpression(*this, e);
            ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xsemantic.hpp:453:37: note: in instantiation of function template specialization
      'xt::xcontainer_semantic<xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 1, xt::layout_type::row_major,
      xt::xtensor_expression_tag> >::assign_xexpression<xt::xreducer<xt::xreducer_functors<(lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), (lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> > >' requested here
        return this->derived_cast().assign_xexpression(e);
                                    ^
/Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xtensor.hpp:384:24: note: in instantiation of function template specialization
      'xt::xsemantic_base<xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 1, xt::layout_type::row_major,
      xt::xtensor_expression_tag> >::assign<xt::xreducer<xt::xreducer_functors<(lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), (lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> > >' requested here
        semantic_base::assign(e);
                       ^
./test.cpp:15:10: note: in instantiation of function template specialization 'xt::xtensor_container<xt::uvector<double,
      std::__1::allocator<double> >, 1, xt::layout_type::row_major,
      xt::xtensor_expression_tag>::xtensor_container<xt::xreducer<xt::xreducer_functors<(lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), (lambda at
      /Users/scheibler/anaconda3/envs/mixiva/include/xtensor/xnorm.hpp:228:5), std::__1::plus<std::__1::complex<double> > >, const
      xt::xtensor_container<xt::uvector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, std::__1::array<unsigned long, 1> > >' requested here
  return xt::norm_sq(a, {0});
         ^
./test.cpp:26:17: note: in instantiation of function template specialization 'mynorm<double>' requested here
  std::cout <<  mynorm(a) << std::endl;
                ^
1 error generated.
```
I'm compiling with Apple LLVM on Mojave (10.14.4). This may be a similar issue to #291.
```
g++ --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/c++/4.2.1
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
Is this expected behavior ?
Thanks in advance!

--- a/include/xtensor/xnorm.hpp
+++ b/include/xtensor/xnorm.hpp
@@ -305,7 +305,8 @@ namespace xt
     template <class T>
     inline auto norm_sq(const std::complex<T>& t) noexcept
     {
-        return std::norm(t);
+        // Does not use std::norm since it returns a std::complex on OSX
+        return t.real() * t.real() + t.imag() * t.imag();
     }
 
     /**
@@ -366,6 +367,24 @@ namespace xt
     }
 #endif
 
+    namespace detail
+    {
+        template <class T>
+        struct norm_value_type
+        {
+            using type = T;
+        };
+
+        template <class T>
+        struct norm_value_type<std::complex<T>>
+        {
+            using type = T;
+        };
+
+        template <class T>
+        using norm_value_type_t = typename norm_value_type<T>::type;
+    }
+
 #define XTENSOR_EMPTY
 #define XTENSOR_COMMA ,
 #define XTENSOR_NORM_FUNCTION(NAME, RESULT_TYPE, REDUCE_EXPR, REDUCE_OP, MERGE_FUNC) \
@@ -374,7 +393,7 @@ namespace xt
     inline auto NAME(E&& e, X&& axes, EVS es = EVS()) noexcept                       \
     {                                                                                \
         using value_type = typename std::decay_t<E>::value_type;                     \
-        using result_type = RESULT_TYPE;                                             \
+        using result_type = detail::norm_value_type_t<RESULT_TYPE>;                  \
                                                                                      \
         auto reduce_func = [](result_type const& r, value_type const& v) {           \
             return REDUCE_EXPR(r REDUCE_OP NAME(v));                                 \

--- a/test/test_xcomplex.cpp
+++ b/test/test_xcomplex.cpp
@@ -13,6 +13,7 @@
 #include "xtensor/xbuilder.hpp"
 #include "xtensor/xcomplex.hpp"
 #include "xtensor/xio.hpp"
+#include "xtensor/xnorm.hpp"
 
 namespace xt
 {

--- a/test/test_xnorm.cpp
+++ b/test/test_xnorm.cpp
@@ -119,14 +119,32 @@ namespace xt
     {
         xarray<std::complex<double>> a = -ones<std::complex<double>>({9});
 
-        EXPECT_EQ(norm_l0(a)(), 9u);
-        EXPECT_EQ(norm_lp_to_p(a, 0.0)(), 9.0);
-        EXPECT_EQ(norm_l1(a)(), 9.0);
-        EXPECT_EQ(norm_lp(a, 1.0)(), 9.0);
-        EXPECT_EQ(norm_sq(a)(), 9.0);
-        EXPECT_EQ(norm_l2(a)(), 3.0);
-        EXPECT_EQ(norm_lp(a, 2.0)(), 3.0);
-        EXPECT_EQ(norm_linf(a)(), 1.0);
+        auto l0_res = norm_l0(a)();
+        auto lp_to_p_res = norm_lp_to_p(a, 0.0)();
+        auto l1_res = norm_l1(a)();
+        auto lp_res1 = norm_lp(a, 1.0)();
+        auto sq_res = norm_sq(a)();
+        auto l2_res = norm_l2(a)();
+        auto lp_res2 = norm_lp(a, 2.0)();
+        auto linf_res = norm_linf(a)();
+
+        EXPECT_TRUE((std::is_same<decltype(l0_res), long long unsigned int>::value));
+        EXPECT_TRUE((std::is_same<decltype(lp_to_p_res), double>::value));
+        EXPECT_TRUE((std::is_same<decltype(l1_res), double>::value));
+        EXPECT_TRUE((std::is_same<decltype(lp_res1), double>::value));
+        EXPECT_TRUE((std::is_same<decltype(sq_res), double>::value));
+        EXPECT_TRUE((std::is_same<decltype(l2_res), double>::value));
+        EXPECT_TRUE((std::is_same<decltype(lp_res2), double>::value));
+        EXPECT_TRUE((std::is_same<decltype(linf_res), double>::value));
+
+        EXPECT_EQ(l0_res, 9u);
+        EXPECT_EQ(lp_to_p_res, 9.0);
+        EXPECT_EQ(l1_res, 9.0);
+        EXPECT_EQ(lp_res1, 9.0);
+        EXPECT_EQ(sq_res, 9.0);
+        EXPECT_EQ(l2_res, 3.0);
+        EXPECT_EQ(lp_res2, 3.0);
+        EXPECT_EQ(linf_res, 1.0);
 
         EXPECT_EQ(norm_l0(a, {0}, evaluation_strategy::immediate)(), 9u);
         EXPECT_EQ(norm_lp_to_p(a, 0.0, {0}, evaluation_strategy::immediate)(), 9.0);

